Secret manager template with project (#2283)

guillaumeblaquiere opened the pull request:
In relation with the issue #2283

This first step is focused on the Secret Manager Template bean. If it's on the right way, I will add the bootstrap part to load secret from another projects

pivotal-issuemaster commented:
@guillaumeblaquiere Please sign the [Contributor License Agreement](https://cla.pivotal.io/sign/spring?repositoryId=spring-cloud/spring-cloud-gcp&pullRequestId=2284)!

[Click here](https://cla.pivotal.io/sync/spring?repositoryId=spring-cloud/spring-cloud-gcp&pullRequestId=2284) to manually synchronize the status of this Pull Request.

See the [FAQ](https://cla.pivotal.io/about) for frequently asked questions.

pivotal-issuemaster commented:
@guillaumeblaquiere Thank you for signing the [Contributor License Agreement](https://cla.pivotal.io/sign/spring?repositoryId=spring-cloud/spring-cloud-gcp&pullRequestId=2284)!

dzou commented:
@guillaumeblaquiere - Thanks for the PR! We will take a look. Also we'll see what the result of our offline discussion is regarding whether we would like to limit the scope of the template to 1 project or allow it to query any project.

meltsufin commented on CONTRIBUTING.adoc:
```diff
@@ -47,7 +47,7 @@ Install the plugin `Eclipse Code Formatter`.
 
 Then navigate to `Settings > Other Settings` (this might be under `Preferences` on Mac) and select the Eclipse Code Formatter.
-Select the `eclipse-code-formatter.xml` file for the field `Eclipse Java Formatter config file` and the file `eclipse.importorder` for the field `Import order`.
+Select the `eclipse-code-formatter.xml` file for the field `Eclipse Java Formatter config file` and the file `eclipse.importorder` for the field `Import order` in the `src/eclipse` directory
```
```suggestion
Select the `eclipse-code-formatter.xml` file for the field `Eclipse Java Formatter config file` and the file `eclipse.importorder` for the field `Import order` in the `src/eclipse` directory.
```

guillaumeblaquiere replied:
done in newer commit

meltsufin commented on spring-cloud-gcp-secretmanager/src/main/java/org/springframework/cloud/gcp/secretmanager/SecretManagerOperations.java:
```diff
@@ -40,6 +40,21 @@
+	 * @param payload the secret payload string.
+	 * @param projectId unique identifier of your project.
+	 */
+	void createSecretWithProject(String secretId, String payload, String projectId);
```
I would just create an overloaded method `createSecret(secretId, payload, projectId)`

guillaumeblaquiere replied:
done in newer commit

dzou commented on spring-cloud-gcp-secretmanager/src/test/java/org/springframework/cloud/gcp/secretmanager/SecretManagerTemplateTests.java:
```diff
@@ -90,6 +90,35 @@ public void testCreateSecretIfMissing() {
 	}
 
+	@Test
+	public void testCreateSecretIfMissingWithProject() {
```
```suggestion
	public void testCreateSecretIfMissing_withProject() {
```

Minor style - if you could make modify the test name to be `_withProject` I think it is more clear that it refers to using the project parameter

guillaumeblaquiere replied:
done in newer commit

guillaumeblaquiere commented:
I'm totally agree with you and I wanted to overload!! But I can't. Indeed, Java method signature are on the type, not on the param name and there were a conflict between these (for example):

```
	@Override
	public String getSecretString(String secretId, String versionName) {....}

	@Override
	public String getSecretString(String secretId, String projectId) {....}
```

Thus, I had to change the method name for differentiating the signatures. And for being consistent in  the whole code, I added, the `WithProject` suffix to all the methods that offer the possibility to set the `ProjectId` param.

Another idea is to require a GcpProjectIdProvider type for projectId insteand of a String
```
	@Override
	public String getSecretString(String secretId, GcpProjectIdProvider projectId) {....}
```

And to let the developer to perform this kind of call
```
	String secret = secretManagerTemplate.getSecretString("my-secret", () -> "custom-project");
```
It works, but, IMO, it's not clean and readable.

Let me know what you prefer.

dzou commented:
> I'm totally agree with you and I wanted to overload!! But I can't. Indeed, Java method signature are on the type, not on the param name and there were a conflict between these (for example):
> 
> ```
> 	@Override
> 	public String getSecretString(String secretId, String versionName) {....}
> 
> 	@Override
> 	public String getSecretString(String secretId, String projectId) {....}
> ```
> 
> Thus, I had to change the method name for differentiating the signatures. And for being consistent in the whole code, I added, the `WithProject` suffix to all the methods that offer the possibility to set the `ProjectId` param.

That is a good point. Maybe we could just omit adding the `String getSecretStringWithProject(String secretId, String projectId)`, and only add `String getSecretString(String secretId, String version, String projectId)`?

I.e. don't add the method that lets users specify only `secretId` and `projectId`. Maybe this is not too bad of a compromise to keep the overloading? I don't think users would mind having to specify the version in addition to the project id.

guillaumeblaquiere commented:
> I.e. don't add the method that lets users specify only secretId and projectId. Maybe this is not too bad of a compromise to keep the overloading? I don't think users would mind having to specify the version in addition to the project id.

Yes, it's an good option. I though about it, didn't know the best solution. Go ahead with yours.

guillaumeblaquiere commented:
Updated with the bootstrap part.

I added a Static public variable for the latestVersion. I did this because the user need to specify the version in case of use of `ProjectId` param. For helping him, the static variable is here to provide right latestVersion string value.

Sometime, the convention is to set in UPPER_CASE the static value. Tell me if it's the case in your convention too.

dzou commented on spring-cloud-gcp-secretmanager/src/main/java/org/springframework/cloud/gcp/secretmanager/SecretManagerTemplate.java:
```diff
@@ -40,6 +40,11 @@
+	/**
+	 * Default value for the latest version of the secret.
+	 */
+	public static final String latestVersion = "latest";
```
Oh yup, I see. We would prefer all uppercase name here: `LATEST_VERSION`.

guillaumeblaquiere replied:
Pushed

dzou commented:
Thanks Guillaume, I'm also gonna help merge in master to your branch and also push some refdoc updates to your PR branch.

guillaumeblaquiere commented:
Coverage result is strange!

dzou commented:
Hey @guillaumeblaquiere - Would you mind if we split this PR and only do the template changes in this one? And then we will address the changes to the property source loading for multiple project IDs in the second one?

Sorry to ask this and make you undo some work. Just wanted to limit the scope of the PR and also give us an opportunity to discuss how to best expose loading secrets from multiple projects as property source.

guillaumeblaquiere commented:
Done (with rebase into master to resolve conflicts).

guillaumeblaquiere commented:
New PR for bootstrap: #2289

dzou commented:
Thanks for the splitting up the PR. Added some minor commits and merged master into the branch. Looks good on my end.

guillaumeblaquiere commented:
Thanks. Travis is broken. Tell me if this PR need more work.
Best

dzou commented:
> Thanks. Travis is broken. Tell me if this PR need more work.
> Best

Just pushed the fix. I think the PR is in a good state; I'll take care of any remaining comments that Mike might have.

meltsufin commented on spring-cloud-gcp-secretmanager/src/main/java/org/springframework/cloud/gcp/secretmanager/SecretManagerOperations.java:
```diff
@@ -73,6 +103,18 @@
+	 * @param projectId unique identifier of your project.
+	 * @return The secret payload as String
+	 */
+	String getSecretString(String secretId, String versionName, String projectId);
```
Are we missing `String getSecretString(String secretId, String projectId)` variation or would this be too much?

guillaumeblaquiere replied:
No, it's not missing. We discussed previously about this.

> Java method signature are on the type, not on the param name and there were a conflict between these (for example):
>
>```
>	@Override
>	public String getSecretString(String secretId, String versionName) {....}
>
>	@Override
>	public String getSecretString(String secretId, String projectId) {....}
>```

So we choose to only propose the projectId with the version. And that's why I also introduced a static public variable `LATEST_VERSION` for helping the developers when they want to get the secret without care about the version

But, we can explore another solution.

meltsufin commented on spring-cloud-gcp-secretmanager/src/main/java/org/springframework/cloud/gcp/secretmanager/SecretManagerOperations.java:
```diff
@@ -111,4 +177,14 @@
+	 * @param projectId unique identifier of your project.
+	 * @return true if the secret exists in Secret Manager; false otherwise
+	 */
+	boolean secretExists(String secretId, String projectId);
```
Does it make sense to have `secretExists(String secretId, String versionName, String projectId)`?
This would check if a specific version of the secret exists.

guillaumeblaquiere replied:
I guess that the primary use of this function is internal for the secret creation (create a new secret or add a version to an existing secret). Thereby, the version is useless in this context.

However, if we want to offer a convenient interface to developer, it could be interesting. If you agree, I propose these interfaces:

```
boolean secretExists(String secretId);// Check latest version in default project
boolean secretExists(String secretId, String version); // Check specific version in default project
boolean secretExists(String secretId, String version, String projectId);// // Check specific version in specific project
```
Like for your `getSecretString` comment, I won't be able to keep `boolean secretExists(String secretId, String projectId);` because of method signature conflict with `boolean secretExists(String secretId, String version);`

Let me know.

dzou replied:
Yeah let's omit adding `secretExists(String secretId, String versionName, String projectId)`, since it would conflict with existing method signature and I think its value is relatively limited as you say with it being primarily internal.

meltsufin replied:
Sorry for the noise, you are absolutely right about the method signature collisions. I didn't think of that.

meltsufin commented:
@guillaumeblaquiere @dzou I'm a little concerned with proliferation of the different method permutations. We have `getSecret`, `getSecretString`, and `getSecretByteString` multiplied by 3 variations of each. Maybe we should think about how common the use of each method will be. I'm thinking that maybe we just need to add the project and version permutations to the `getSecretByteString` because it's pretty trivial to go from `ByteString` to `byte[]` and `String`. We get the benefits of the flexibility of being able to specify the project and the version without adding too much to the cognitive load of the users of the template. You can think of the users of the `getSecretByteString` as the power users, and the users of the version and project are also power users. So, we can take care of their needs with just two additional methods.

Similarly, for the `createSecret` methods we should probably just choose one type of payload, probably `byte[]` to focus on to give the version and project flexibility on.

What do you guys think?

guillaumeblaquiere commented:
@meltsufin, Your concern is correct. I think that your solution acceptable:

* Less methods -> easier to maintain and read
* For "exceptional" use, the template offer the capability to specify the projectId but the developer have to do few "extra" work for getting his secret (in string or byteString).

It's a good balance.

@dzou, what do you think?

dzou commented:
> @meltsufin, Your concern is correct. I think that your solution acceptable:
> 
> * Less methods -> easier to maintain and read
> * For "exceptional" use, the template offer the capability to specify the projectId but the developer have to do few "extra" work for getting his secret (in string or byteString).
> 
> It's a good balance.
> 
> @dzou, what do you think?

Sounds good to me. Just pushed the changes.

meltsufin commented on spring-cloud-gcp-samples/spring-cloud-gcp-secretmanager-sample/src/main/java/com/example/SecretManagerWebController.java:
```diff
@@ -46,25 +48,58 @@
+
+		String secretPayload;
+		if (StringUtils.isEmpty(projectId)) {
+			secretPayload = this.secretManagerTemplate.getSecretString(secretId, version);
```
Should we remove this variant of `getSecretString` as well? Same goes for `getSecretBytes`.

guillaumeblaquiere replied:
These variants exists in the version 1.2.2. We shouldn't remove them to ensure the downward compatibility.

meltsufin replied:
Ah, didn't realize that. Thanks!
